flink-yarn/src/main/java/org/apache/flink/yarn/cli/FlinkYarnSessionCli.java

```diff
@@ -260,8 +263,15 @@ public AbstractYarnClusterDescriptor createDescriptor(String defaultApplicationN
 		} else {
 			LOG.info("No path for the flink jar passed. Using the location of "
 				+ yarnClusterDescriptor.getClass() + " to locate the jar");
-			localJarPath = new Path("file://" +
-				yarnClusterDescriptor.getClass().getProtectionDomain().getCodeSource().getLocation().getPath());
+			String encodedJarPath =
+				yarnClusterDescriptor.getClass().getProtectionDomain().getCodeSource().getLocation().getPath();
+			try {
+				// we have to decode the url encoded parts of the path
+				localJarPath = new Path("file://" + URLDecoder.decode(encodedJarPath, Charset.defaultCharset().name()));
```

Comment:
I think manually coding the `file:// + path` URIs have cross-platform issues.

For File URIs, I found it to be best to go with 
```java
new File(string).toURI()
```
or in that specific case:
```java
localJarPath = new Path(new File(URLDecoder.decode(encodedJarPath, Charset.defaultCharset().name())).toURI());
or
localJarPath = new Path(new URI("file", null, URLDecoder.decode(encodedJarPath, Charset.defaultCharset().name()), null));
```